Convert the lobby to a React app

davideaster opened the pull request:
This PR converts the lobby from the server-rendered Jade templates to a browser-rendered React app.

There are three outstanding issues:

  - [x] The Scenarios tab doesn't render correctly. 
  - [x] *Sessions* is incorrectly set as the default tab for instructors. The default should be *Scenarios*.
  - [x] The Logout button isn't active.

#-1 occurs because each row contains a `form` and forms aren't valid in a `tr`. This inadvertently worked in plain Bootstrap, but it isn't styled correctly in `react-bootstrap`. The fix is to replace the table with Bootstrap rows and columns. However, the upcoming switch to a sortable, filterable table view will require similar but different changes. It may not be worth fixing the table just to immediately replace it with something else.

#-2 is caused by the delay in loading the user data. The built-in user that's applied to the first render should be marked as an instructor. This way, the `Scenarios` tab exists in the first render and the default can be applied.

#-3 is a mystery since it worked previously but there have been no apparent code changes. This needs investigation.

There are a lot of changes. To make review as easy as possible the commits are arranged as a steady progression from a React shell app to the functioning lobby.

Changes can be grouped as:

## React *Hello, world!*

  - `package.json, shrinkwrap, .gitignore`
  - `ESLint`
  - `Mocha`
  - `JSDoc`
  - `webpack`
  - `Configure webpack to bundle CSS`
  - `Configure webpack for production and development builds`
  - `Babel`
  - `React`
  - `Bootstrap`
  - `Enable Babel Stage 2 extensions to allow class fields`
  - `Relax ESLint errors`

## Static layout

Transcribe static content from the Jade templates into React components. The layout uses `react-bootstrap`, which allows the page to be expressed using a slightly higher abstraction than native Bootstrap and removes some noisiness from the Bootstrap layout. `react-bootstrap` also uses React mechanisms for Bootstrap components and removes the dependency on jQuery.

  - `Layout`
  - `Scenarios, Sessions, and Review components with placeholder content`

## Mock locals

Add a temporary `locals` module to stand in for local variables provided to the Jade templates. The locals consist of utility functions, application version data, user data, and the server manifest.

  - `Add functions previously provided to the Jade templates as locals`
  - `Add mock version and user locals previously provided to the Jade template`
  - `Add mock manifest locals`

## `locals` rendering

Update the layout to be informed by the `locals` data. Render pane content.

  - `Configure the layout using version and user data`
  - `Render scenario, session and review content using the manifest data`

## React controls

Connect form inputs to component state. Derive the button show/hide and enable/disable settings from the state.

  - ``Convert `Application` to a class and maintain inputs in the state``
  - ``Convert `Scenario` to a class and maintain input values in the state``

## Form submission

Draw data for form submission from the state.

  - `Add jQuery to support form submission`
  - `Submit new-application and new-scenario forms`
  - `Submit the import form`

## Mock data in state

Connect `locals` data to component state. Render from state instead of directly from the locals. Move `locals` functions to the components.

  - ``Convert `Layout` to a class and render version and user data from the state``
  - `Render scenario, session and review content from the layout state`

## jQuery to `fetch`

Replace `$.post` with `fetch`.

  - ``Add `get` and `post` wrappers for `fetch` ``
  - ``Switch jQuery POST operations to `fetch` via `post` ``
  - `Remove jQuery`

## Live data

Add services to expose the former `locals` data. Load component state from the services.

  - ``Add `version` and `user` services for data previously available as Jade locals``
  - `Load version and user data from the corresponding services`
  - ``Add a `manifest` service``
  - `Load manifest data from the service`
  - `Remove mock locals`

## Refresh

Reload the manifest and re-render instead of reloading the page after posting new data to the server.

  - `Refresh the manifest after posting data to the server`

## Login page

  - `Login page`

## Integration

Build and serve the React lobby. Remove the Jade lobby.

  - ``Build the lobby pages on `npm install` ``
  - `Replace the Jade index and login pages with the React versions`
  - `Remove the Jade lobby`

davideaster committed:
Render scenario, session and review content using the manifest data

davideaster committed:
Add functions previously provided to the Jade templates as locals

davideaster committed:
Enable Babel Stage 2 extensions to allow class fields

davideaster committed:
Convert `Scenario` to a class and maintain input values in the state

davideaster committed:
Convert `Application` to a class and maintain inputs in the state

davideaster committed:
Submit new-application and new-scenario forms

davideaster committed:
Submit the import form

davideaster committed:
Convert `Layout` to a class and render version and user data from the state

davideaster committed:
Add jQuery to support form submission

davideaster committed:
Render scenario, session and review content from the layout state

davideaster committed:
Load manifest data from the service

davideaster committed:
Add a `manifest` service

davideaster committed:
Switch jQuery POST operations to `fetch` via `post`

davideaster committed:
Remove jQuery

davideaster committed:
Refresh the manifest after posting data to the server

davideaster committed:
Login page

davideaster committed:
Add `version` and `user` services for data previously available as Jade locals

davideaster committed:
Add `get` and `post` wrappers for `fetch`

davideaster committed:
Replace the Jade index and login pages with the React versions

davideaster committed:
Add mock version and user locals previously provided to the Jade template

davideaster committed:
Add mock manifest locals

davideaster committed:
Load version and user data from the corresponding services

davideaster committed:
Configure the layout using version and user data

davideaster committed:
Remove mock locals

davideaster committed:
Layout

davideaster committed:
JSDoc

davideaster committed:
Configure webpack to bundle CSS

davideaster committed:
React

davideaster committed:
Babel

davideaster committed:
webpack

davideaster committed:
Configure webpack for production and development builds

davideaster committed:
package.json, shrinkwrap, .gitignore

davideaster committed:
Scenarios, Sessions, and Review components with placeholder content

davideaster committed:
Bootstrap

davideaster committed:
ESLint

davideaster committed:
Mocha

davideaster committed:
Build the lobby pages on `npm install`

davideaster committed:
Relax ESLint errors

davideaster committed:
Remove the Jade lobby

eric79 commented:
Awesome, awesome, awesome.  But when I run the app, the lobby doesn't look like it used to:

![image](https://user-images.githubusercontent.com/3160246/34953409-b8dc065a-f9ea-11e7-9843-56108d30cd47.png)

I see that I'm getting a handful of errors in the console:

![image](https://user-images.githubusercontent.com/3160246/34956246-47d38954-f9f6-11e7-8f2e-e507b89e6561.png)

eric79 commented on support/lobby/index.html:
```diff
@@ -0,0 +1,21 @@
+  <link rel="stylesheet" href="node_modules/bootstrap/dist/css/bootstrap-theme.min.css">
+  <link rel="stylesheet" href="index.bundle.css">
+
+</head><body>
```
Interesting indentation scheme ... I would normally put `<body>` on its own line to signal the beginning of a new block.  But it looks like you've adopted more of an `Egyptian brackets` style.  I don't have strong opinions either way.  Just thought that I would comment.  :smile:

davideaster replied:
Since there's almost always a `head` and a `body` and nothing else at the top, I don't see any need to preserve additional indention within `html` . It looks odd if you're looking for the strict tree, but I like that this way clearly identifies and separates the two sections.

davideaster commented:
I added belated comments about some outstanding issues to the description. Please see the notes about the Scenarios tab there.

eric79 commented on support/lobby/lib/Review.js:
```diff
@@ -0,0 +1,58 @@
+      <Head/>
+    </thead>
+    <tbody>
+      <Reviewws records={ props.records }/>
```
I have to admit that I'm not a fan of the misspellings of `Reviewws` and `Scenarioos` to reflect an inner element ... might I suggest a name like `ReviewList` or `ReviewRows` or some such?

davideaster replied:
Yes, this is some silliness I should have cleaned up. It was originally `Scenarios` since it's the set of `Scenario` elements, but I had to change it to make way for the containing `Scenarios` component. The component could have been unnamed, but the React debugger wouldn't identify the component in that case.

eric79 commented on support/lobby/lib/utils.js:
```diff
@@ -0,0 +1,34 @@
+}
+
+function postOptions( body ) {
+  let json = typeof body === "object" && body !== null &&
```
I wouldn't mind some parentheses here to visually delimit the conditions from the conjunctions.

eric79 commented:
Are you planning to address the above-mentioned three issues on this PR?

davideaster commented:
#-2 and #-3, yes. I don't want to spend time on #-1 just to immediately redo it when switching to a React table. We shouldn't merge it with #-1 unaddressed, of course.

davideaster committed:
Defer the first render until service data is available

davideaster committed:
Fix logout form

davideaster committed:
Fix row component names

davideaster committed:
Add parens for clarity

davideaster committed:
`npm install --save react-table`

youngca reviewed (commented):
Ditto to Eric's review.  The layout needs to look closer to (or better than) the original layout.  I'm seeing the same error messages in the console.

BTW, I rebased this branch on tdg on 2/7/2018.

davideaster committed:
Correct capitalization on Scenarios, Sessions, and Review imports

davideaster committed:
`react-table` component to allow rows to manage state for cells

davideaster committed:
`react-table` CSS

davideaster committed:
Scenarios (but not yet Application) as `react-table`

davideaster committed:
Sessions as `react-table`

davideaster committed:
Review as `react-table`

davideaster committed:
Enable filters

davideaster committed:
Disable sorting on non-data columns

davideaster committed:
Case-insensitive, partial string filtering

davideaster committed:
Fix filtering on the custom Sessions title column

davideaster committed:
Replace form `onSubmit` with button `onClick` to remove form from row

davideaster committed:
Use link formatting for link buttons

davideaster committed:
Remove unused `react-bootstrap` imports

davideaster committed:
Allow unused function arguments in ESLint

davideaster committed:
Remove unnecessary `instructorStudentsLabel` export

davideaster committed:
Add IDs to distinguish columns

davideaster committed:
app row

davideaster committed:
Render the Import label as a button

davideaster committed:
Switch the Export link to a button

davideaster committed:
Restore the view refresh after creating a session

eric79 committed:
Change fetch header from null to undefined to avoid type error

eric79 commented on support/client/lib/version.js:
```diff
@@ -30,7 +30,7 @@ define( function() {
     /// bumping the version number.
 
-    var version = [ 0, 8, 0, "", "", [ "ITDG", 2, 8, 2 ] ];  // version-identifier
+    var version = [ 0, 8, 0, "", "", [ "ITDG", 2, 8, 6 ] ];  // version-identifier
```
Do you remember why we have this in two places? ... I know that I asked this last time ... I always forget the answer.  :wink:

youngca committed:
Reduce timeout from 15s to 5s.

At 15s, the screen stays blank for about 15s on a lobby refresh.  That's not a good user experience.
Too long and a user is going to repeatedly hit the refresh button in frustration.
At a 5s timeout, the screen is blank for 1-5s, far more tolerable for a user and within the
recommended range.

eric79 committed:
Rename lobby tabs

eric79 committed:
Change Active Sessions tab to only show Active Sessions

eric79 committed:
Modify rules for displaying sessions as Active or Previous

Active Sessions:
- Instance exists
- Is marked as !isReview

Previous Sessions:
- No instance exists

eric79 committed:
Ensure that Date column starts wide enough to show whole date

eric79 committed:
Center the Review and Resume links in their table cell

eric79 committed:
Line up text and links on lobby Review tab rows

eric79 committed:
Rename Sessions to ActiveSessions and Review to PreviousSessions

eric79 committed:
Refactor record functions to just return what the caller needs

eric79 committed:
Explicitly display the ActiveSessions TabPane for students

eric79 committed:
Hide filter inputs until at least one scenario exists

eric79 committed:
Break super long line into several in Scenarios.js

eric79 committed:
Undo hack to make student TabContent visible

eric79 committed:
Remove extra ;

eric79 committed:
Reduce Evaluate timeout to 1s

youngca commented:
:+1:
